File: CulinarySiteApp/src/app/services/dish.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { IDishListModel } from '../interfaces/dish/dish-list-model.interface';
import { IDishDetailModel } from '../interfaces/dish/dish-detail-model.interface';
import { CreateDishModel } from '../viewmodels/dish/create-dish-model.class';
import { UpdateDishModel } from '../viewmodels/dish/update-dish-model.class';
import { IDishModel } from '../interfaces/dish/dish-model.interface';

@Injectable()
export class DishService {
  public url: string = '/api/dish';

  constructor(private http: HttpClient) {}

  public getDishDetailList(withRelated: boolean): Observable<IDishListModel[]> {
    return this.http.get<IDishListModel[]>(this.url + '/' + withRelated);
  }

  public getDishList(): Observable<IDishModel[]> {
    return this.http.get<IDishModel[]>(this.url);
  }

  public getDish(
    id: number,
    withRelated: boolean
  ): Observable<IDishDetailModel> {
    return this.http.get<IDishDetailModel>(
      this.url + '/' + id + '/' + withRelated
    );
  }

  public createDish(createDishModel: CreateDishModel): Observable<void> {
    return this.http.post<void>(this.url, createDishModel);
  }

  public updateDish(updateDishModel: UpdateDishModel): Observable<void> {
    return this.http.put<void>(this.url, updateDishModel);
  }

  public deleteDish(id: number): Observable<void> {
    return this.http.delete<void>(this.url + '/' + id);
  }
}

File: CulinarySite/CulinarySite.Bll/Services/DishService.cs
using System.Collections.Generic;
using AutoMapper;
using CulinarySite.Bll.Interfaces;
using CulinarySite.Common.Dtos.Dish;
using CulinarySite.Dal.Interfaces;
using CulinarySite.Domain.Entities;
using System.Linq;
using CulinarySite.Common.Exceptions;

namespace CulinarySite.Bll.Services
{
    public class DishService : IDishService
    {
        private readonly IReadOnlyGenericRepository<Dish> _dishReadOnlyRepository;
        private readonly IWriteGenericRepository<Dish> _dishWriteRepository;
        private readonly IMapper _mapper;
        public DishService(
            IReadOnlyGenericRepository<Dish> dishReadOnlyRepository,
            IWriteGenericRepository<Dish> dishWriteRepository,
            IMapper mapper)
        {
            _dishReadOnlyRepository = dishReadOnlyRepository;
            _dishWriteRepository = dishWriteRepository;
            _mapper = mapper;
        }

        public void CreateDish(CreateDishDto createDishDto)
        {
            var dishCategories = _dishReadOnlyRepository.GetItemList().Select(x => x.Category);

            if (dishCategories.Contains(createDishDto.Category))
            {
                throw new ValidationException($"The dish category:{createDishDto.Category} already exists.");
            }

            Dish dish = _mapper.Map<Dish>(createDishDto);

            _dishWriteRepository.Create(dish);
            _dishWriteRepository.Save();
        }

        public void UpdateDish(UpdateDishDto updateDishDto)
        {
            Dish dish = _mapper.Map<Dish>(updateDishDto);

            _dishWriteRepository.Update(dish);
            _dishWriteRepository.Save();
        }

        public void DeleteDish(int id)
        {
            _dishWriteRepository.Delete(id);
            _dishWriteRepository.Save();
        }

        public DishDetailDto GetDish(int id, bool withRelated)
        {
            var dish = new Dish();
            var dishDetailDto = new DishDetailDto();

            if (withRelated)
            {
                dish = _dishReadOnlyRepository.GetItemWithInclude(
                                x => x.Id == id,
                                x => x.Recipes
                                );

                dishDetailDto = _mapper.Map<DishDetailDto>(dish);

                return dishDetailDto;
            }

            dish = _dishReadOnlyRepository.GetItem(id);
            dishDetailDto = _mapper.Map<DishDetailDto>(dish);

            return dishDetailDto;
        }

        public IEnumerable<DishListDto> GetDishDetailList(bool withRelated)
        {
            IEnumerable<Dish> dishes;
            IEnumerable<DishListDto> dishListDtos;

            if (withRelated)
            {
                dishes = _dishReadOnlyRepository.GetItemListWithInclude(x => x.Recipes);

                dishListDtos = dishes.Select(x => _mapper.Map<DishListDto>(x));

                return dishListDtos;
            }

            dishes = _dishReadOnlyRepository.GetItemList();

            dishListDtos = dishes.Select(x => _mapper.Map<DishListDto>(x));

            return dishListDtos;
        }

        public IEnumerable<DishDto> GetDishList()
        {
            IEnumerable<Dish> dishes = _dishReadOnlyRepository.GetItemList();
            var dishDtos = dishes.Select(x => _mapper.Map<DishDto>(x));

            return dishDtos;
        }
    }
}

